File: android/feature/scripture/src/main/kotlin/com/nexussoft/verbum/feature/scripture/AudioPlayerFeature.kt
package com.nexussoft.verbum.feature.scripture

import com.nexussoft.verbum.clients.AudioPlayerClient
import com.nexussoft.verbum.clients.AudioPlayerEvent
import com.nexussoft.verbum.clients.NowPlayingInfo
import com.nexussoft.verbum.clients.helloao.ScriptureAudioClient
import com.nexussoft.verbum.common.arch.Effect
import com.nexussoft.verbum.common.arch.Reducer
import com.nexussoft.verbum.common.arch.only
import com.nexussoft.verbum.common.arch.runEffect
import com.nexussoft.verbum.common.arch.with
import com.nexussoft.verbum.models.AudioNarrator
import com.nexussoft.verbum.models.ChapterAudio
import com.nexussoft.verbum.models.PassageReference
import kotlinx.coroutines.ensureActive

/**
 * The one player in the app. Lives on `AppFeature`, so listening survives navigation;
 * the mini player renders it anywhere. Chapters chain. Twin of iOS `AudioPlayerFeature`.
 */
object AudioPlayerFeature {
    data class State(
        val reference: PassageReference? = null,
        val audio: ChapterAudio? = null,
        val narrator: AudioNarrator? = null,
        val isPlaying: Boolean = false,
        val isLoading: Boolean = false,
        val currentTime: Double = 0.0,
        val duration: Double = 0.0,
        val rate: Float = 1f,
        val failed: Boolean = false,
    ) {
        val isActive: Boolean get() = reference != null
        val progress: Float get() = if (duration > 0) (currentTime / duration).coerceAtMost(1.0).toFloat() else 0f
        fun isPlaying(reference: PassageReference) = isPlaying && this.reference?.bookId == reference.bookId && this.reference?.chapter == reference.chapter
    }

    sealed interface Action {
        data class Play(val reference: PassageReference) : Action
        data class AudioLoaded(val audio: ChapterAudio?) : Action
        data object TogglePlayPause : Action
        data object SkipForward : Action
        data object SkipBackward : Action
        data class Seek(val seconds: Double) : Action
        data object RateTapped : Action
        data class NarratorSelected(val narrator: AudioNarrator) : Action
        data object StopTapped : Action
        data object ChapterTapped : Action
        data class Event(val event: AudioPlayerEvent) : Action
        data class Delegate(val delegate: DelegateAction) : Action
    }

    sealed interface DelegateAction {
        data class OpenChapter(val reference: PassageReference) : DelegateAction
    }

    const val SKIP_SECONDS = 15.0
    val RATES = listOf(1f, 1.25f, 1.5f, 0.8f)

    private object LoadId
    private object PlayerId

    fun reducer(audioClient: ScriptureAudioClient, player: AudioPlayerClient): Reducer<State, Action> = Reducer { state, action ->
        when (action) {
            is Action.Play -> {
                val chapter = PassageReference(action.reference.bookId, action.reference.chapter)
                State(reference = chapter, narrator = state.narrator, rate = state.rate, isLoading = true).with(
                    Effect.Merge(listOf(
                        runEffect(id = PlayerId, cancelInFlight = true) {},
                        runEffect(id = LoadId, cancelInFlight = true) { send ->
                            player.stop()
                            val audio = try { audioClient.chapterAudio(chapter.bookId, chapter.chapter) }
                            catch (e: kotlinx.coroutines.CancellationException) { throw e }
                            catch (_: Exception) { null }
                            kotlinx.coroutines.currentCoroutineContext().ensureActive()
                            send(Action.AudioLoaded(audio))
                        },
                    )),
                )
            }

            is Action.AudioLoaded -> {
                val audio = action.audio
                if (state.reference == null || (audio != null && audio.reference != state.reference)) state.only()
                else if (audio == null || audio.narrators.isEmpty()) state.copy(isLoading = false, failed = true).only()
                else {
                    val narrator = audio.narrators.firstOrNull { it.id == state.narrator?.id } ?: audio.narrators[0]
                    val next = state.copy(audio = audio, narrator = narrator)
                    next.with(start(next, player))
                }
            }

            is Action.NarratorSelected ->
                if (action.narrator == state.narrator) state.only()
                else state.copy(narrator = action.narrator, currentTime = 0.0, isLoading = true).let { it.with(start(it, player)) }

            Action.TogglePlayPause ->
                if (state.audio == null || state.isLoading || state.failed) state.only()
                else state.with(runEffect { if (state.isPlaying) player.pause() else player.play() })

            Action.SkipForward -> state.with(Effect.Send(Action.Seek(minOf(state.duration, state.currentTime + SKIP_SECONDS))))
            Action.SkipBackward -> state.with(Effect.Send(Action.Seek(maxOf(0.0, state.currentTime - SKIP_SECONDS))))

            is Action.Seek ->
                if (!state.isActive) state.only()
                else state.copy(currentTime = action.seconds).with(runEffect { player.seek(action.seconds) })

            Action.RateTapped -> {
                val rate = RATES[(RATES.indexOf(state.rate).coerceAtLeast(0) + 1) % RATES.size]
                state.copy(rate = rate).with(runEffect { player.setRate(rate) })
            }

            Action.StopTapped -> State().with(
                Effect.Merge(listOf(
                    runEffect(id = PlayerId, cancelInFlight = true) {},
                    runEffect(id = LoadId, cancelInFlight = true) {},
                    runEffect { player.stop() },
                )),
            )

            Action.ChapterTapped -> state.reference?.let { state.with(Effect.Send(Action.Delegate(DelegateAction.OpenChapter(it)))) } ?: state.only()

            is Action.Event -> if (state.audio == null) state.only() else when (val e = action.event) {
                is AudioPlayerEvent.Ready -> state.copy(duration = e.durationSeconds, isLoading = false).only()
                is AudioPlayerEvent.Time -> state.copy(currentTime = e.seconds).only()
                is AudioPlayerEvent.Playing -> if (e.isPlaying == state.isPlaying) state.only() else state.copy(isPlaying = e.isPlaying).only()
                AudioPlayerEvent.Ended -> {
                    val next = state.reference?.let(ChapterNavigation::next)
                    if (next == null) state.copy(isPlaying = false).only()
                    else state.copy(isPlaying = false).with(Effect.Send(Action.Play(next)))
                }
                AudioPlayerEvent.Failed -> state.copy(isLoading = false, isPlaying = false, failed = true).only()
            }

            is Action.Delegate -> state.only()
        }
    }

    /** Load the recording, start it, and keep listening to the player until stopped or replaced. */
    private fun start(state: State, player: AudioPlayerClient): Effect<Action> {
        val narrator = state.narrator ?: return Effect.Send(Action.Event(AudioPlayerEvent.Failed))
        val audio = state.audio ?: return Effect.Send(Action.Event(AudioPlayerEvent.Failed))
        val reference = state.reference ?: return Effect.Send(Action.Event(AudioPlayerEvent.Failed))
        val info = NowPlayingInfo(reference.formatted, "${audio.translationName} · ${narrator.name}")
        val rate = state.rate
        return runEffect(id = PlayerId, cancelInFlight = true) { send ->
            player.load(narrator.url, info)
            player.setRate(rate)
            player.play()
            player.events.collect { send(Action.Event(it)) }
        }
    }
}

File: VerbumKit/Sources/Features/Audio/AudioPlayerFeature.swift
import Clients
import ComposableArchitecture
import Foundation
import Models

/// The one player in the app. Lives on `AppFeature`, not on a chapter, so
/// listening survives navigation; the mini player renders it anywhere.
/// Chapters chain: when one ends, the next begins.
@Reducer
public struct AudioPlayerFeature {
    @ObservableState
    public struct State: Equatable {
        public var reference: PassageReference?
        public var audio: ChapterAudio?
        public var narrator: AudioNarrator?
        public var isPlaying = false
        public var isLoading = false
        public var currentTime: TimeInterval = 0
        public var duration: TimeInterval = 0
        public var rate: Float = 1
        public var failed = false

        public init() {}

        public var readingPosition: AudioReadingPosition? {
            guard !failed, !isLoading, let audio, let cue = AudioCue.active(in: narrator?.cues ?? [], at: currentTime) else { return nil }
            return .init(reference: audio.reference, translationID: audio.translationId, cue: cue, isPlaying: isPlaying)
        }
        public var isActive: Bool { reference != nil }
        public var progress: Double { duration > 0 ? min(1, currentTime / duration) : 0 }

        public func isPlaying(_ reference: PassageReference) -> Bool {
            isPlaying && self.reference?.bookId == reference.bookId && self.reference?.chapter == reference.chapter
        }
    }

    public enum Action: Equatable {
        /// Start listening to a chapter (from the reader, or the next one after the last ended).
        case play(PassageReference)
        case audioResponse(ChapterAudio?)
        case togglePlayPause
        case skipForward
        case skipBackward
        case seek(TimeInterval)
        case rateTapped
        case narratorSelected(AudioNarrator)
        case stopTapped
        case chapterTapped
        case event(AudioPlayerEvent)
        case delegate(Delegate)

        @CasePathable
        public enum Delegate: Equatable {
            case openChapter(PassageReference)
        }
    }

    @Dependency(\.scriptureAudio) var scriptureAudio
    @Dependency(\.audioPlayer) var player

    public init() {}

    static let skip: TimeInterval = 15
    static let rates: [Float] = [1, 1.25, 1.5, 0.8]

    public var body: some ReducerOf<Self> {
        Reduce { state, action in
            switch action {
            case .play(let reference):
                let chapter = PassageReference(bookId: reference.bookId, chapter: reference.chapter)
                state.reference = chapter
                state.audio = nil
                state.isLoading = true
                state.failed = false
                state.isPlaying = false
                state.currentTime = 0
                state.duration = 0
                return .merge(
                    .cancel(id: CancelID.player),
                    .run { [scriptureAudio, player] send in
                        await player.stop()
                        let audio = try? await scriptureAudio.chapterAudio(bookId: chapter.bookId, chapter: chapter.chapter)
                        try Task.checkCancellation()
                        await send(.audioResponse(audio))
                    }
                    .cancellable(id: CancelID.load, cancelInFlight: true)
                )

            case .audioResponse(let audio):
                guard let reference = state.reference else { return .none }
                if let audio, audio.reference != reference { return .none }
                guard let audio, !audio.narrators.isEmpty else {
                    state.isLoading = false
                    state.failed = true
                    return .none
                }
                state.audio = audio
                // Keep the narrator the listener chose if this chapter has them too.
                let narrator = audio.narrators.first { $0.id == state.narrator?.id } ?? audio.narrators[0]
                state.narrator = narrator
                return start(narrator, state: state)

            case .narratorSelected(let narrator):
                guard narrator != state.narrator else { return .none }
                state.narrator = narrator
                state.currentTime = 0
                state.isLoading = true
                return start(narrator, state: state)

            case .togglePlayPause:
                guard state.audio != nil, !state.isLoading, !state.failed else { return .none }
                let playing = state.isPlaying
                return .run { [player] _ in playing ? await player.pause() : await player.play() }

            case .skipForward:
                return .send(.seek(min(state.duration, state.currentTime + Self.skip)))

            case .skipBackward:
                return .send(.seek(max(0, state.currentTime - Self.skip)))

            case .seek(let time):
                guard state.isActive else { return .none }
                state.currentTime = time
                return .merge(
                    .run { [player] _ in await player.seek(to: time) },
                    nowPlaying(state)
                )

            case .rateTapped:
                let index = Self.rates.firstIndex(of: state.rate) ?? 0
                state.rate = Self.rates[(index + 1) % Self.rates.count]
                let rate = state.rate
                return .merge(
                    .run { [player] _ in await player.setRate(rate) },
                    nowPlaying(state)
                )

            case .stopTapped:
                state = State()
                return .merge(
                    .cancel(id: CancelID.player),
                    .cancel(id: CancelID.load),
                    .run { [player] _ in await player.stop() }
                )

            case .chapterTapped:
                guard let reference = state.reference else { return .none }
                return .send(.delegate(.openChapter(reference)))

            case .event(.ready(let duration)):
                guard state.audio != nil else { return .none }
                state.duration = duration
                state.isLoading = false
                return nowPlaying(state)

            case .event(.time(let time)):
                guard state.audio != nil else { return .none }
                let previous = state.currentTime
                state.currentTime = time
                if Int(previous / 10) != Int(time / 10) { return nowPlaying(state) }
                return .none

            case .event(.playing(let playing)):
                guard state.audio != nil else { return .none }
                guard playing != state.isPlaying else { return .none }
                state.isPlaying = playing
                return nowPlaying(state)

            case .event(.ended):
                guard state.audio != nil else { return .none }
                state.isPlaying = false
                guard let reference = state.reference, let next = ChapterNavigation.next(after: reference) else {
                    return .none
                }
                return .send(.play(next))

            case .event(.failed):
                guard state.audio != nil else { return .none }
                state.isLoading = false
                state.isPlaying = false
                state.failed = true
                return .none

            case .event(.remote(let command)):
                switch command {
                case .play: return .run { [player] _ in await player.play() }
                case .pause: return .run { [player] _ in await player.pause() }
                case .togglePlayPause: return .send(.togglePlayPause)
                case .skipForward: return .send(.skipForward)
                case .skipBackward: return .send(.skipBackward)
                case .seek(let time): return .send(.seek(time))
                }

            case .delegate:
                return .none
            }
        }
    }

    /// Load the recording, start it, and keep listening to the player until stopped or replaced.
    private func start(_ narrator: AudioNarrator, state: State) -> Effect<Action> {
        guard let url = URL(string: narrator.url) else { return .send(.event(.failed)) }
        let rate = state.rate
        return .run { [player] send in
            await player.load(url)
            await player.setRate(rate)
            let events = player.events()
            await player.play()
            for await event in events {
                await send(.event(event))
            }
        }
        .cancellable(id: CancelID.player, cancelInFlight: true)
    }

    private func nowPlaying(_ state: State) -> Effect<Action> {
        guard let reference = state.reference, let audio = state.audio, let narrator = state.narrator else { return .none }
        let info = NowPlayingInfo(
            title: reference.formatted,
            subtitle: "\(audio.translationName) · \(narrator.name)",
            elapsed: state.currentTime,
            duration: state.duration,
            rate: state.isPlaying ? state.rate : 0
        )
        return .run { [player] _ in await player.updateNowPlaying(info) }
    }

    private enum CancelID { case load, player }
}
